internal/azure/exchanger.go

```diff
@@ -0,0 +1,67 @@
+package azure
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+)
+
+type tokenResponse struct {
+	AccessToken  string `json:"access_token"`
+	RefreshToken string `json:"refresh_token"`
+	Resource     string `json:"resource"`
+	TokenType    string `json:"token_type"`
+}
+
+type Exchanger struct {
+	acrFQDN string
+}
+
+func NewExchanger(acrEndpoint string) *Exchanger {
+	return &Exchanger{
+		acrFQDN: acrEndpoint,
+	}
+}
+
+func (e *Exchanger) ExchangeACRAccessToken(armToken string) (string, error) {
+	exchangeUrl := fmt.Sprintf("%s://%s/oauth2/exchange", "https", e.acrFQDN)
+	parsedURL, err := url.Parse(exchangeUrl)
+	if err != nil {
+		return "", err
+	}
+
+	parameters := url.Values{}
+	parameters.Add("grant_type", "access_token")
+	parameters.Add("service", parsedURL.Hostname())
+	parameters.Add("access_token", armToken)
+
+	resp, err := http.PostForm(exchangeUrl, parameters)
+	if err != nil {
+		return "", fmt.Errorf("failed to send token exchnage request: %w", err)
+	}
+
+	if resp.StatusCode != 200 {
+		responseBytes, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+
+		}
+
+		return "", fmt.Errorf("unexpected status code %d from exchnage request, body: %s",
+			resp.StatusCode, string(responseBytes))
+	}
+
+	responseBytes, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return "", fmt.Errorf("failed to read response body: %w", err)
+	}
+
+	var tokenResp tokenResponse
+	err = json.Unmarshal(responseBytes, &tokenResp)
```

Comment:
`ioutil.ReadAll` loads everything into memory and it should be avoided whenever possible. Using the json Decoder would be more efficient.
```suggestion
        var tokenResp tokenResponse
        decoder := json.NewDecoder(resp.Body)
        if err := decoder.Decode(&tokenResp); err != nil {
        	return "", err
        }
```